core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/ICEExplainer.scala

```diff
@@ -0,0 +1,252 @@
+
+  val averageKind = "average"
+  val individualKind = "individual"
+
+  val categoricalFeatures = new TypedArrayParam[ICECategoricalFeature] (
+    this,
+    "categoricalFeatures",
+    "The list of categorical features to explain.",
+    _.forall(_.validate)
+  )
+
+  def setCategoricalFeatures(values: Seq[ICECategoricalFeature]): this.type = this.set(categoricalFeatures, values)
+  def getCategoricalFeatures: Seq[ICECategoricalFeature] = $(categoricalFeatures)
+
+  val numericFeatures = new TypedArrayParam[ICENumericFeature] (
+    this,
+    "numericFeatures",
+    "The list of numeric features to explain.",
+    _.forall(_.validate)
+  )
+
+  def setNumericFeatures(values: Seq[ICENumericFeature]): this.type = this.set(numericFeatures, values)
+  def getNumericFeatures: Seq[ICENumericFeature] = $(numericFeatures)
+
+  val kind = new Param[String] (
+    this,
+    "kind",
+    "Whether to return the partial dependence plot (PDP) averaged across all the samples in the " +
+      "dataset or individual feature importance (ICE) per sample. " +
+      "Allowed values are \"average\" for PDP and \"individual\" for ICE.",
+    ParamValidators.inArray(Array(averageKind, individualKind))
+  )
+
+  def getKind: String = $(kind)
+  def setKind(value: String): this.type = set(kind, value)
+
+  setDefault(kind -> "individual", numericFeatures -> Seq.empty[ICENumericFeature],
+    categoricalFeatures -> Seq.empty[ICECategoricalFeature])
+}
+
+/**
+  * ICETransformer displays the model dependence on specified features with the given dataframe
+  * as background dataset. It supports 2 types of plots: individual - dependence per instance and
+  * average - across all the samples in the dataset.
+  * Note: This transformer only supports one-way dependence plot.
+  */
+@org.apache.spark.annotation.Experimental
+class ICETransformer(override val uid: String) extends Transformer
+  with HasExplainTarget
+  with HasModel
+  with ICEFeatureParams
+  with Wrappable
+  with ComplexParamsWritable {
+
+  override protected lazy val pyInternalWrapper = true
+
+  def this() = {
+    this(Identifiable.randomUID("ICETransformer"))
+  }
+
+  private def calcDependence(df: DataFrame, idCol: String, targetClassesColumn: String,
+                             feature: String, values: Array[_], outputColName: String): DataFrame = {
+
+    val dataType = df.schema(feature).dataType
+    val explodeFunc = explode(array(values.map(v => lit(v)): _*).cast(ArrayType(dataType)))
+
+    val predicted = getModel.transform(df.withColumn(feature, explodeFunc))
+    val targetCol = DatasetExtensions.findUnusedColumnName("target", predicted)
+    val dependenceCol = DatasetExtensions.findUnusedColumnName("feature__dependence", predicted)
+
+    val explainTarget = extractTarget(predicted.schema, targetClassesColumn)
+    val result = predicted.withColumn(targetCol, explainTarget)
+
+    getKind.toLowerCase match {
+      case this.averageKind =>
```

Comment:
nit: idt you need the "this" here

Reply:
Added backticks, because it's looking for a stable identifier. Ref: https://stackoverflow.com/questions/7078022/why-does-pattern-matching-in-scala-not-work-with-variables